File: golang/serializers/xml_serializer.go
// serializers/xml_serializer.go
package serializers

import (
	"encoding/xml"
	"golang/models"
	"time"
)

type XMLSerializer struct{}

func (s *XMLSerializer) Serialize(user *models.User) ([]byte, error) {
	start := time.Now()
	data, err := xml.Marshal(user)
	_ = time.Since(start) // Можно сохранить для метрик
	return data, err
}

func (s *XMLSerializer) Deserialize(data []byte) (*models.User, error) {
	var user models.User
	err := xml.Unmarshal(data, &user)
	return &user, err
}

func (s *XMLSerializer) MeasureSerializationTime(user *models.User, iterations int) time.Duration {
	var totalTime time.Duration

	for i := 0; i < iterations; i++ {
		start := time.Now()
		xml.Marshal(user)
		totalTime += time.Since(start)
	}

	return totalTime / time.Duration(iterations)
}

func (s *XMLSerializer) MeasureDeserializationTime(data []byte, iterations int) time.Duration {
	var totalTime time.Duration

	for i := 0; i < iterations; i++ {
		start := time.Now()
		var user models.User
		xml.Unmarshal(data, &user)
		totalTime += time.Since(start)
	}

	return totalTime / time.Duration(iterations)
}

File: java/src/main/java/ram/ka/ru/formats/XmlSerializer.java
// XmlSerializer.java
package ram.ka.ru.formats;

import ram.ka.ru.models.User;

import javax.xml.bind.JAXBContext;
import javax.xml.bind.JAXBException;
import javax.xml.bind.Marshaller;
import javax.xml.bind.Unmarshaller;
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;

public class XmlSerializer {
    private final JAXBContext jaxbContext;
    
    public XmlSerializer() {
        try {
            jaxbContext = JAXBContext.newInstance(User.class);
        } catch (JAXBException e) {
            throw new RuntimeException("Failed to create JAXB context", e);
        }
    }
    
    public byte[] serialize(User user) throws IOException {
        try {
            Marshaller marshaller = jaxbContext.createMarshaller();
            marshaller.setProperty(Marshaller.JAXB_FORMATTED_OUTPUT, Boolean.FALSE);
            marshaller.setProperty(Marshaller.JAXB_FRAGMENT, Boolean.TRUE);
            
            ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
            marshaller.marshal(user, outputStream);
            return outputStream.toByteArray();
        } catch (JAXBException e) {
            throw new IOException("XML serialization failed", e);
        }
    }
    
    public User deserialize(byte[] data) throws IOException {
        try {
            Unmarshaller unmarshaller = jaxbContext.createUnmarshaller();
            ByteArrayInputStream inputStream = new ByteArrayInputStream(data);
            return (User) unmarshaller.unmarshal(inputStream);
        } catch (JAXBException e) {
            throw new IOException("XML deserialization failed", e);
        }
    }
    
    public long measureSerializationTime(User user, int iterations) throws IOException {
        long totalTime = 0;
        for (int i = 0; i < iterations; i++) {
            long startTime = System.nanoTime();
            serialize(user);
            long endTime = System.nanoTime();
            totalTime += (endTime - startTime);
        }
        return totalTime / iterations;
    }
}